rest-service/manager_rest/storage/filters.py

```diff
@@ -0,0 +1,130 @@
+import re
+
+from manager_rest import manager_exceptions
+from manager_rest.storage.models_base import db
+
+
+def add_labels_filters_to_query(query, labels_model, labels_filters):
+    query = query.join(labels_model)
+    for labels_filter in labels_filters:
+        try:
+            if '!=' in labels_filter:
+                label_key, raw_label_value = labels_filter.split('!=')
+                label_value = _get_label_value(raw_label_value)
+                if isinstance(label_value, list):
+                    query = query.filter(key_not_equal_list_values(
+                        labels_model, label_key, label_value))
+                else:
+                    query = query.filter(key_not_equal_value(
+                        labels_model, label_key, label_value))
+
+            elif '=' in labels_filter:
+                label_key, raw_label_value = labels_filter.split('=')
+                label_value = _get_label_value(raw_label_value)
+                if isinstance(label_value, list):
+                    query = query.filter(key_equal_list_values(
+                        labels_model, label_key, label_value))
+                else:
+                    query = query.filter(
+                        key_equal_value(labels_model, label_key, label_value))
+
+            elif 'null' in labels_filter:
+                match_null = re.match(r'\S+ is null', labels_filter)
+                match_not_null = re.match(r'\S+ is not null', labels_filter)
+                if match_null:
+                    query = query.filter(
+                        key_not_exist(labels_model, match_null.group(1)))
+                elif match_not_null:
+                    query = query.filter(
+                        key_exist(labels_model, match_not_null.group(1)))
+                else:
+                    _raise_bad_labels_filter(labels_filter)
+
+            else:
+                _raise_bad_labels_filter(labels_filter)
+
+        except SyntaxError:
+            _raise_bad_labels_filter(labels_filter)
+
+    return query
+
+
+def _raise_bad_labels_filter_value(labels_filter_value):
+    raise manager_exceptions.BadParametersError(
+        'The labels filter value `{0}` must be a string or a list '
+        'of strings'.format(labels_filter_value)
+    )
+
+
+def _raise_bad_labels_filter(labels_filter_value):
+    raise manager_exceptions.BadParametersError(
+        'The labels filter `{0}` is not in the right '
+        'format. It must be one of: <key>=<value>, '
+        '<key>=[<value1>,<value2>,...], <key>!=<value>, '
+        '<key>!=[<value1>,<value2>,...], <key> is null, '
+        '<key> is not null'.format(labels_filter_value)
+    )
+
+
+def _get_label_value(raw_label_value):
+    # raw_label_value should be of the form [<value1>,<value2>,...]
+    # in order to count as a list
+    match_list = re.match(r'^\[(\S+)\]$', raw_label_value)
```

Comment:
This means that this list `[b, c]` is invalid, due to the space.

Reply:
Shouldn't this be r'^\[([^\]]+)\]$'?

Reply:
Right, this makes more sense, changing.